File: 2024/jl/day05.jl
# Day 5: Print Queue

include("util.jl")

using .Util
using DataStructures: DefaultDict

r, p = split.(strip.(split(Util.getinputn("day05.txt"), "\n\n")), "\n")
rules = DefaultDict(Set)
for rule ∈ r
	f, t = split(rule, "|")
	push!(rules[f], t)
end

function validate()
	valid, invalid = [], []

	for pages ∈ p
		visited = Set()
        good = true

		pages = split(pages, ",")
        for page ∈ pages
			if !isdisjoint(rules[page], visited)
                good = false
                break
			end

            push!(visited, page)
		end

		if good push!(valid, pages) else push!(invalid, pages) end
	end

	return valid, invalid
end

function star1()
	valid, _ = validate()

	total = sum(parse(Int64, x[length(x) ÷ 2 + 1]) for x ∈ valid)
	println("- $total")
end

function star2()
	_, invalid = validate()

	compare(a, b) = a ∈ rules[b]
	sort!.(invalid, lt=compare)

	total = sum(parse(Int64, x[length(x) ÷ 2 + 1]) for x ∈ invalid)
	println("- $total")
end

star1()
star2()

File: 2024/py/src/day05.py
# Day 5: Print Queue

from src import util

rules, pages = util.input_single("day05.txt").split("\n\n")

rules = rules.split("\n")
pages = pages.split("\n")


def star1():
    rule = {}
    for r in rules:
        f, t = r.split("|")

        if f in rule:
            rule[f].add(t)
        else:
            rule[f] = {t}

    valid = []
    for page in pages:
        visited = set()
        good = True
        page = page.split(",")

        for p in page:
            if (p in rule) and not rule[p].isdisjoint(visited):
                good = False
                break

            visited.add(p)

        if good:
            valid.append(page)

    return sum(int(x[(len(x) - 1) // 2]) for x in valid)


def star2():
    rule = {}
    for r in rules:
        f, t = r.split("|")

        if f in rule:
            rule[f].add(t)
        else:
            rule[f] = {t}

    invalid = []
    for page in pages:
        visited = set()
        bad = False
        page = page.split(",")

        for p in page:
            if (p in rule) and not rule[p].isdisjoint(visited):
                bad = True
                break

            visited.add(p)

        if bad:
            invalid.append(page)

    for v in invalid:
        swaps = 0
        while True:
            for i in range(len(v) - 1):
                if (v[i + 1] in rule) and v[i] in rule[v[i + 1]]:
                    v[i], v[i + 1] = v[i + 1], v[i]
                    swaps += 1

            if swaps == 0:
                break

            swaps = 0

    return sum(int(x[(len(x) - 1) // 2]) for x in invalid)


util.run_day(5, star1, star2)
